constraint-solver/src/algebraic_constraint/solve.rs

```diff
@@ -164,15 +164,14 @@ where
         &self,
         range_constraints: &impl RangeConstraintProvider<T::FieldType, V>,
     ) -> Result<ProcessResult<T, V>, Error> {
-        let (_, mut linear, constant) = self.expression.components();
-
-        Ok(if linear.clone().count() == 1 {
-            let (var, coeff) = linear.next().unwrap();
+        Ok(if self.expression.linear_components().count() == 1 {
+            let (var, coeff) = self.expression.linear_components().next().unwrap();
```

Comment:
This could be lazier, now it goes through the whole iterator to get the count, where it could get the first element and branch on whether there's a second one. Non blocking